### PR TITLE
Fix: improve bracket parsing error, set Slice type to Unknown

### LINKED ISSUE #2437: Unexpected `AttributeError` from `parse_one`
Running the following raises `AttributeError: 'NoneType' object has no attribute 'type'`

```py
import sqlglot

sql = 'Ā[:'

sqlglot.parse_one(sql, dialect='postgres')
```

I would expect a `ParseError` (and indeed that's what I get when using `transpile`).

Goes with out saying that this was again picked out by [hypothesis](https://hypothesis.readthedocs.io/en/latest/).

I'm using sqlglot version: 18.16.1 on Python 3.11.6.

<details>

<summary>Traceback:</summary>

```
(env311) code/pydantic-data-platform 1 ➤ python test.py
Traceback (most recent call last):
  File ".../test.py", line 5, in <module>
    sqlglot.parse_one(sql, dialect='postgres')
  File ".../env311/lib/python3.11/site-packages/sqlglot/__init__.py", line 125, in parse_one
    result = dialect.parse(sql, **opts)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/dialects/dialect.py", line 311, in parse
    return self.parser(**opts).parse(self.tokenize(sql), sql)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 985, in parse
    return self._parse(
           ^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 1051, in _parse
    expressions.append(parse_method(self))
                       ^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 1240, in _parse_statement
    expression = self._parse_expression()
                 ^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3174, in _parse_expression
    return self._parse_alias(self._parse_conjunction())
                             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3177, in _parse_conjunction
    return self._parse_tokens(self._parse_equality, self.CONJUNCTION)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 4813, in _parse_tokens
    this = parse_method()
           ^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3180, in _parse_equality
    return self._parse_tokens(self._parse_comparison, self.EQUALITY)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 4813, in _parse_tokens
    this = parse_method()
           ^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3183, in _parse_comparison
    return self._parse_tokens(self._parse_range, self.COMPARISON)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 4813, in _parse_tokens
    this = parse_method()
           ^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3186, in _parse_range
    this = self._parse_bitwise()
           ^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3293, in _parse_bitwise
    this = self._parse_term()
           ^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3318, in _parse_term
    return self._parse_tokens(self._parse_factor, self.TERM)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 4813, in _parse_tokens
    this = parse_method()
           ^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/dialects/postgres.py", line 346, in _parse_factor
    return self._parse_tokens(self._parse_exponent, self.FACTOR)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 4813, in _parse_tokens
    this = parse_method()
           ^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/dialects/postgres.py", line 349, in _parse_exponent
    return self._parse_tokens(self._parse_unary, self.EXPONENT)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 4813, in _parse_tokens
    this = parse_method()
           ^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3326, in _parse_unary
    return self._parse_at_time_zone(self._parse_type())
                                    ^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3335, in _parse_type
    this = self._parse_column()
           ^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3520, in _parse_column
    return self._parse_column_ops(this)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 3523, in _parse_column_ops
    this = self._parse_bracket(this)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/parser.py", line 4059, in _parse_bracket
    expressions = apply_index_offset(this, expressions, -self.INDEX_OFFSET)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/helper.py", line 184, in apply_index_offset
    annotate_types(expression)
  File ".../env311/lib/python3.11/site-packages/sqlglot/optimizer/annotate_types.py", line 56, in annotate_types
    return TypeAnnotator(schema, annotators, coerces_to).annotate(expression)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/optimizer/annotate_types.py", line 383, in annotate
    return self._maybe_annotate(expression)  # This takes care of non-traversable expressions
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/optimizer/annotate_types.py", line 392, in _maybe_annotate
    annotator(self, expression)
  File ".../env311/lib/python3.11/site-packages/sqlglot/optimizer/annotate_types.py", line 262, in <lambda>
    expr_type: lambda self, e: self._annotate_binary(e)
                               ^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../env311/lib/python3.11/site-packages/sqlglot/optimizer/annotate_types.py", line 431, in _annotate_binary
    left_type, right_type = left.type.this, right.type.this
                            ^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'type'
```

</details>

### PR DESCRIPTION
**Pull request by GeorgeSittas**
Fixes #2437

### CODE REVIEW

File: sqlglot/optimizer/annotate_types.py [hunk 1]
```diff
@@ -287,6 +287,7 @@ class TypeAnnotator(metaclass=_TypeAnnotator):
         exp.Max: lambda self, e: self._annotate_by_args(e, "this", "expressions"),
         exp.Min: lambda self, e: self._annotate_by_args(e, "this", "expressions"),
         exp.Null: lambda self, e: self._annotate_with_type(e, exp.DataType.Type.NULL),
+        exp.Slice: lambda self, e: self._annotate_with_type(e, exp.DataType.Type.UNKNOWN),
```
Comment (tobymao) [inline sqlglot/optimizer/annotate_types.py:290]: shouldn’t alice be i’m the t of list <t>

Comment (GeorgeSittas) [reply sqlglot/optimizer/annotate_types.py:290]: Ah due to unknown

Comment (tobymao) [reply sqlglot/optimizer/annotate_types.py:290]: can’t slice type be inferred if the underlying obj is an list<t>

Comment (GeorgeSittas) [reply sqlglot/optimizer/annotate_types.py:290]: I think slice corresponds to the "range" object and the bracket is what you mean? Not sure if it makes sense to assign a type to 1:3 for example, but a[1:3] can have eg a type of array of ints.

Comment (tobymao) [reply sqlglot/optimizer/annotate_types.py:290]: ah ok

**Review by tobymao — APPROVED** [review]

File: sqlglot/optimizer/annotate_types.py (hunk 1 above)
Comment (GeorgeSittas) [reply sqlglot/optimizer/annotate_types.py:290]: I'm testing the subarray type we discussed now, might need to make a few small changes to fix it too

**Pull request merged**